### PR TITLE
Todo routes

### PR DESCRIPTION
**Pull request by Abiodun-Shittu**
### Changes To Be Made By This PR
- set up various endpoints for the todo routes
- GET, POST, PATCH and DELETE

### PR Review Steps
- [ ] import and export of necessary libraries
- [ ] set up endpoints for various routes to read, create, update and delete todos
- [ ] Approve

### CODE REVIEW

File: src/modules/todos/controllers.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+	};
+};
+
+export function getTodo(req, res) {
+	const id = req.params.id;
+	const findTodo = todos.find((todo) => todo.id === id);
+	if (!findTodo) {
+		throw new AppException(404, "Unable to retrieve todo")
+	}
+	return res.json({
```
Comment (IyiKuyoro) [inline src/modules/todos/controllers.js]: This does not seem to follow the response schema.

File: src/modules/todos/controllers.js [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	if (title) {
+		updateTodo.title = title;
+	}
+	if (status) {
+		updateTodo.status = status;
+	}
+	if (date) {
+		updateTodo.date = date;
+	}
+	return res.json({
```
Comment (IyiKuyoro) [inline src/modules/todos/controllers.js]: Again. Response schema mismatch

File: src/modules/todos/controllers.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+};
+
+export function deleteTodo(req, res) {
+	const id = req.params.id;
+	const deleteTodo = todos.find((todo) => todo.id === id);
+	if (!deleteTodo) {
+		throw new AppException(404, "Unable to retrieve todo")
+	};
+	todos = todos.filter((todo) => todo.id !== id);
+	res.json(todos);
```
Comment (IyiKuyoro) [inline src/modules/todos/controllers.js]: The correct response when deleting a resource should be 204 no content. 204 is used for more than just delete. See more about 204 [here](https://www.rfc-editor.org/rfc/rfc7231#section-6.3.5)

**Review by IyiKuyoro — COMMENTED** [review]
You got the structure right but I have a few concerns.

Ideally, a Todo should be created by a user. Todos cannot just exist on their own. This means you would need to find a way to associate the todo with the creating user. You might consider adding the user id to the todo body.

- Only authenticated users should be able to create, read, update or delete (CRUD) todos.
- If two users exist, one owns todo 1 and the other owns todo 2, If user 1 tries to access todo 2, he should get a (403: Unauthorized) or a (404: Not Found).
- When a user tries GET `/todos`, the application should return only the todo's owned by that user. The same goes for updating and deleting. Else you are opening the application up to hackers.

One more thing, where are the tests?

File: src/modules/todos/controllers.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+};
+
+export function createTodo(req, res, next) {
+	try {
+		const title = req.body.title;
+		const status = req.body.status;
+		const date = req.body.date;
+		const todo = {
+			id: req.userId,
```
Comment (IyiKuyoro) [inline src/modules/todos/controllers.js]: looks to me like you are setting the id of the todo to the user id. This seems off. How then would we uniquely identify a todo? May I suggest this instead?
```
{
    id: <something unique for each todo>,
    status: <an enumeration of possible status (pending, started, completed). should always start at pending>,
    title: <the name of the todo>,
    date: <the deadline>,
    userId: <the id of the owner of this todo>,
    createdAt: <the date the todo was created>,
    updatedAt: <the last time the todo was updated>,
}
```

File: src/modules/todos/controllers.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+export function deleteTodo(req, res) {
+	const id = req.params.id;
+	const deleteTodo = todos.find((todo) => todo.id === id);
+	if (!deleteTodo) {
+		throw new AppException(404, "Unable to retrieve todo")
+	};
+	return res.status(204).json({
+		statusCode: 204,
+		message: "Todo successfully deleted"
```
Comment (IyiKuyoro) [inline src/modules/todos/controllers.js]: How? where is the line that does the deleting?

Comment (Abiodun-Shittu) [reply src/modules/todos/controllers.js]: must have mistakenly removed that, will put it back...  
my bad

File: src/modules/todos/controllers.js [hunk 6]
```diff
@@ -40,19 +61,32 @@ export function getTodo(req, res) {
 	return res.status(200).json({
 		statusCode: 200,
 		data: {
+			userId: findTodo.userId,
```
Comment (IyiKuyoro) [inline src/modules/todos/controllers.js]: Please include the id of the todo as well. This function should also ensure that the todo is created by the logged in user.

File: src/modules/todos/controllers.js [hunk 7]
```diff
@@ -0,0 +1,128 @@
+import { todos, users } from '../database/database.js';
+import AppException from '../../utils/exceptions/AppException.js';
+import { v4 } from 'uuid';
+
+
+export function getTodos(_, res) {
+	res.json(todos);
```
Comment (IyiKuyoro) [inline src/modules/todos/controllers.js]:
```
	res.json(todos.filter(todo => todo.userId === req.userId);
```

**Review by IyiKuyoro — APPROVED** [review]

**Pull request merged**